be/src/exprs/vectorized/array_functions.cpp

```diff
@@ -660,6 +660,271 @@ class ArrayContainsImpl {
+            nullable_element = nullable->has_null();
+            elements = nullable->has_null() ? elements : nullable->data_column().get();
+        }
+        if (auto nullable = dynamic_cast<const NullableColumn*>(targets); nullable != nullptr) {
+            nullable_target = nullable->has_null();
+            targets = nullable->has_null() ? targets : nullable->data_column().get();
+        }
+        if (targets->only_null() && !nullable_element) {
+            auto result = UInt8Column::create();
+            result->resize(array.size());
+            return result;
+        }
+        // Expand Only-Null column.
+        if (targets->only_null()) {
+            auto data = down_cast<const NullableColumn*>(elements)->data_column()->clone_empty();
+            targets_holder = NullableColumn::create(std::move(data), NullColumn::create());
+            (void)targets_holder->append_nulls(array.size());
+            targets = targets_holder.get();
+            nullable_target = true;
+        }
+        if (targets->is_constant()) {
+            const_target = true;
+        }
+
+        CHECK(!(const_target && nullable_target));
+
+        if (nullable_element && nullable_target) {
+            return _array_has<true, true, false>(*elements, array_offsets, *targets, target_offsets);
+        } else if (nullable_element) {
+            return const_target ? _array_has<true, false, true>(*elements, array_offsets, *targets, target_offsets)
+                                : _array_has<true, false, false>(*elements, array_offsets, *targets, target_offsets);
+        } else if (nullable_target) {
+            return _array_has<false, true, false>(*elements, array_offsets, *targets, target_offsets);
+        } else {
+            return const_target ? _array_has<false, false, true>(*elements, array_offsets, *targets, target_offsets)
+                                : _array_has<false, false, false>(*elements, array_offsets, *targets, target_offsets);
+        }
+    }
+
+    static const NullColumnPtr merge_nullcolum(const NullableColumn* a, const NullableColumn* b) {
+        if (a && b == nullptr) {
+            return a->null_column();
+        }
+        if (b && a == nullptr) {
+            return b->null_column();
+        }
+        DCHECK_EQ(a->size(), b->size());
+        return FunctionHelper::union_null_column(a->null_column(), b->null_column());
+    }
+
+    static ColumnPtr _array_has_generic(const Column& array, const Column& target) {
+        // has_any(NULL, xxx) | has_any(xxx, NULL) -> NULL
+        if (array.only_null() || target.only_null()) {
+            auto result = NullableColumn::create(Int8Column::create(), NullColumn::create());
+            result->append_nulls(array.size());
+            return result;
+        }
+        DCHECK_EQ(array.size(), target.size());
+        auto array_col = dynamic_cast<const ArrayColumn*>(&array);
+        auto target_col = dynamic_cast<const ArrayColumn*>(&target);
```

Comment:
Suggest to use `gutil/casts/down_cast`. It uses `static_cast` for release mode, and checks the result of `dynmic_cast` non-null for debug mode.